### PR TITLE
first solution

### PR DESCRIPTION
**Pull request by NikitaKarbazin**
[DEMO LINK](https://nikitakarbazin.github.io/react_clock/)

### CODE REVIEW

File: src/helper/FetchName.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const FetchName = async () => {
```
Comment (jstmpelowycz) [inline src/helper/FetchName.js:1]: it's not a class, just a function
```suggestion
const fetchName = async () => {
```

File: src/helper/FetchName.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  const response = await fetch('https://mate.academy/students-api/goods',
+    { cache: 'no-cache' });
+  let name = '';
+
+  await response.json().then(date => {
+    name = date[Math.ceil(Math.random() * date.length)];
+  });
+
+  return name?.name ? name.name : 'Unknown Fruit';
```
Comment (jstmpelowycz) [inline src/helper/FetchName.js:10]:
```suggestion
  return name?.name || 'Unknown Fruit';
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,27 +1,70 @@
...
+            if (this.state.isClockVisible) {
+              this.setState({
+                isClockVisible: false,
+              });
+            }
+
+            return this.setState({
+              isClockVisible: false,
+            });
+          }}
```
Comment (jstmpelowycz) [inline src/App.tsx:55]: you'd rather create a function

File: src/App.tsx [hunk 4]
```diff
@@ -1,27 +1,70 @@
...
+            if (!this.state.isClockVisible) {
+              this.setState({
+                isClockVisible: true,
+              });
+            }
+
+            return this.setState({
+              isClockVisible: true,
+            });
+          }}
```
Comment (jstmpelowycz) [inline src/App.tsx:39]: you'd rather create a function

File: src/helper/FetchName.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+const FetchName = async () => {
+  const response = await fetch('https://mate.academy/students-api/goods',
+    { cache: 'no-cache' });
+  let name = '';
```
Comment (jstmpelowycz) [inline src/helper/FetchName.js:4]: perhaps, another name should be put here?

**Review by jstmpelowycz — CHANGES_REQUESTED** [review]

File: src/components/Clock.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    clearInterval(this.timerId);
+  }
+
+  render() {
+    return (
+      <div className="Clock" data-cy="time">
+        <h1>
+          React clock
+          {' '}
+          {this.props.name ? this.props.name : 'Unknown Fruit'}
```
Comment (illiarerun) [inline src/components/Clock.tsx:49]:
```suggestion
          {this.props.name || 'Unknown Fruit'}
```

File: src/components/Clock.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export class Clock extends React.Component<Props, State> {
+  state = {
+    date: new Date(),
+  };
+
+  constructor(props:Props, public timerId: NodeJS.Timer) {
+    super(props);
+  }
```
Comment (illiarerun) [inline src/components/Clock.tsx:19]: pretty much redundant constructor, just put `timerId` declaration outside
```suggestion
```

File: src/components/Clock.tsx (hunk 6 above)
Comment (illiarerun) [inline src/components/Clock.tsx:49]: Use interpolation here

Comment (illiarerun) [reply src/components/Clock.tsx:49]: and also I'd suggest creating variable for this value

File: src/components/Clock.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+          React clock
+          {' '}
+          {this.props.name ? this.props.name : 'Unknown Fruit'}
+        </h1>
+        <p>
+          Current time:
+          {' '}
+          {this.props.isClockVisible
+            ? this.state.date.toLocaleTimeString()
+            : ''}
```
Comment (illiarerun) [inline src/components/Clock.tsx:56]:
```suggestion
          {this.props.isClockVisible && this.state.date.toLocaleTimeString()}
```

Comment (illiarerun) [reply src/components/Clock.tsx:56]: Also, use destructuring - `this.props` or `this.state` is too much

Comment (illiarerun) [reply src/components/Clock.tsx:56]: Also put it in the variable, it's not okay to keep a lot of logic in `jsx`

Comment (illiarerun) [inline src/components/Clock.tsx:56]: USe interpolation here, also

File: src/App.tsx [hunk 9]
```diff
@@ -1,27 +1,72 @@
...
-};
+class App extends React.Component {
+  state = {
+    clockName: 'FRUIT',
+    isClockVisible: true,
+  };
+
+  changeNameHandler = async () => {
+    this.setState({
+      clockName: await fetchName(),
```
Comment (illiarerun) [inline src/App.tsx:14]: why are you fetching names all the time? just fetch them once, in `componentDidMount` and save all the values in state

File: src/App.tsx [hunk 10]
```diff
@@ -1,27 +1,72 @@
...
+    return (
+      <div>
+        <Clock
+          name={this.state.clockName}
+          isClockVisible={this.state.isClockVisible}
+        />
+        <button
+          type="button"
+          onClick={this.changeToggleStateTrue}
+        >
```
Comment (illiarerun) [inline src/App.tsx:52]: I still can click on `Show Clock` button when clock is shown, make it disabled based on state

File: src/App.tsx [hunk 11]
```diff
@@ -1,27 +1,72 @@
...
+          Show Clock
+        </button>
+        <button
+          type="button"
+          onClick={this.changeToggleStateFalse}
+        >
+          Hide Clock
+        </button>
```
Comment (illiarerun) [inline src/App.tsx:60]: Same here

**Review by illiarerun — CHANGES_REQUESTED** [review]

**Pull request closed without merging**